nodejs/device/src/index.ts

```diff
@@ -379,85 +435,153 @@ class DeviceApp {
+  private sendTelemetry(): [Promise<void>, () => void] {
+    logger.info("Starting telemetry sending.");
+    let cancel: () => void;
+    const promise = new Promise<void>((_, reject) => {
+      let cleanedUp = false;
+      let cleanupAndReject = (err: Error) => {
+        cleanedUp = true;
+        clearInterval(interval);
+        reject(err);
+      };
+      cancel = () => {
+        if (cleanedUp) {
+          return;
+        }
+        logger.warn("Telemetry sending canceled.");
+        cleanupAndReject(new Canceled());
+      };
+      const interval = setInterval(() => {
+        const serviceAckId = uuidv4();
+        const now = new Date();
+        this.sessionMetrics.lastUpdateTimeUtc = now.toISOString();
+        this.sessionMetrics.runTime = msToDurationString(
+          now.getTime() - this.runStart.getTime()
         );
-      })
-      .finally(() => {
-        this.serviceAckWaitList.add(serviceAckId);
-        if (
-          this.serviceAckWaitList.size > settings.sendMessageAllowedFailureCount
-        ) {
-          this.stopDeviceApp(
-            new Error(
-              `The number of service acks being waited on exceeds sendMessageAllowedFailureCount of ${settings.sendMessageAllowedFailureCount}.`
-            )
-          );
+        const message = new Message(
+          JSON.stringify({
+            thief: {
+              cmd: "serviceAckRequest",
+              serviceInstanceId: this.serviceInstanceId,
+              runId: runId,
+              serviceAckId: serviceAckId,
+              serviceAckType: "telemetry",
+              sessionMetrics: this.sessionMetrics,
+            },
+          })
+        );
+        logger.info(`Sending message with serviceAckId: ${serviceAckId}`);
+        this.client
+          .sendEvent(message)
+          .catch(() => {
+            logger.error(
+              `Error sending message with serviceAckId: ${serviceAckId}`
+            );
+          })
+          .finally(() => {
+            this.serviceAckWaitList.add(serviceAckId);
+            if (
+              this.serviceAckWaitList.size >
+              settings.sendMessageAllowedFailureCount
+            ) {
+              cleanupAndReject(
+                new Error(
+                  `The number of service acks being waited on exceeds sendMessageAllowedFailureCount of ${settings.sendMessageAllowedFailureCount}.`
+                )
+              );
+            }
+          });
+      }, 1000 * settings.sendMessageOperationsPerSecond);
+    });
+    return [promise, cancel];
+  }
+
+  private sendThiefReportedProperties(): [Promise<void>, () => void] {
+    logger.info("Starting thief reported property sending.");
+    let cancel: () => void;
+    const promise = new Promise<void>((_, reject) => {
+      let cleanedUp = false;
+      let cleanUpAndReject = (err: Error) => {
+        cleanedUp = true;
+        clearInterval(interval);
+        reject(err);
+      };
+      cancel = () => {
```

Comment:
>cancel [](start = 6, length = 6)

exitReason is one of the reported properties.  The python code pushes reported properties one last time after cancel so the exit reason can be pushed up.  Depending on the failure, the push might not succeed, but it almost always does (in python at least).